src/crypto/index.js

```diff
@@ -281,42 +315,29 @@ function _maybeUploadOneTimeKeys(crypto) {
 
         // Ask olm to generate new one time keys, then upload them to synapse.
         crypto._olmDevice.generateOneTimeKeys(keysThisLoop);
-        return _uploadOneTimeKeys(crypto).then(() => {
+        return _uploadOneTimeKeys(crypto).then((res) => {
+            const keyCount = res.one_time_key_counts.signed_curve25519 || 0;
+            if (keyCount != 0) {
+                return uploadLoop(Math.max(keyLimit - keyCount, 0));
+            }
             return uploadLoop(numberToGenerate - keysThisLoop);
         });
     }
 
     crypto._oneTimeKeyCheckInProgress = true;
     Promise.resolve().then(() => {
+        if (crypto._currentKeyCount !== undefined) {
+            // this._currentKeyCount got set and the handling got postponed.
```

Comment:
This isn't clear, something like this would be better:

"We already have the current key count from a /sync response. Use this value instead of asking the server for the current key count."